test/integration/grpc_retry_test.cpp

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gtest/gtest.h>
+
+#include <grpcpp/server_context.h>
+#include <grpcpp/support/status.h>
+#include "framework/test_logger.hpp"
+#include "network/impl/grpc_channel_builder.hpp"
+#include "qry_responses.pb.h"
+#include "queries.pb.h"
+#include "torii/query_service.hpp"
+#include "main/server_runner.hpp"
+
+class MockQueryService : public iroha::torii::QueryService {
+ public:
+  explicit MockQueryService(int max_attempts)
+      : iroha::torii::QueryService(
+          nullptr, nullptr, nullptr, getTestLogger("MockQueryService")),
+        max_attempts_(max_attempts) {}
+
+  grpc::Status Find(grpc::ServerContext *context,
+                    const iroha::protocol::Query *request,
+                    iroha::protocol::QueryResponse *response) override {
+    if (attempts_ < max_attempts_) {
+      attempts_++;
+      return grpc::Status(grpc::StatusCode::UNAVAILABLE, "");
+    }
+    return grpc::Status::OK;
+  }
+
+ private:
+  int attempts_ = 0;
+  int max_attempts_;
+};
+
+namespace {
+  constexpr int kAttemptsForSuccess = 3;  // 5 max attempts at reaching
+  constexpr int kAttemptsForFailure = 5;  // 5 max attempts at reaching
+  constexpr auto kListenIP = "127.0.0.1";
+  constexpr auto kListenPort = 31337;
+
+  auto makeRunner() {
+    auto listen_addr =
+        std::string(kListenIP) + ":" + std::to_string(kListenPort);
+    auto logger = getTestLogger("TestServerRunner");
+    return std::make_shared<ServerRunner>(listen_addr, logger, true);
+  }
+
+  std::shared_ptr<ServerRunner> makeServer(int max_attempts) {
+    auto runner = makeRunner();
+    runner->append(std::make_shared<MockQueryService>(max_attempts));
+    runner->run();
+    return runner;
+  }
+
+  auto makeRequestAndCheckStatus(grpc::StatusCode code) {
+    auto client =
+        iroha::network::createClient<iroha::protocol::QueryService_v1>(
+            std::string(kListenIP) + ":" + std::to_string(kListenPort));
+
+    iroha::protocol::Query query;
+    iroha::protocol::QueryResponse response;
+
+    grpc::ClientContext client_context;
+
+    auto status = client->Find(&client_context, query, &response);
+
+    SCOPED_TRACE(status.error_code());
+    SCOPED_TRACE(status.error_message());
+    SCOPED_TRACE(status.error_details());
```

Comment:
Why `SCOPED_TRACE`s? I see no subroutines. You can print a nice message with `<<` to ASSERT_EQ.

Reply:
Removed these in the latest version, they were not particularly useful anyway